### ISSUE TITLE
Missing "focus" event for "rofi" app

### ISSUE THREAD

**Issue by Hi-Angel** [swaywm/sway#2859]
When `rofi -show run` is executed, its window is shown and focused, e.g. it accepts input from keyboard. However its appearance does not produce "focus" event.

I found it as part of working on [script for transparency](https://github.com/swaywm/sway/issues/2787).

# Steps to reproduce

1. Install from system package manager or pip `i3ipc-python` *(a python library for i3-protocol IPC)*.
2. Install `rofi`
3. Save the following as `test.py`, and execute `python ./test.py`:
	```python
	import i3ipc

	ipc = i3ipc.Connection()

	def on_window_focus(unused, focused):
		print('focused id=' + str(focused.container.id))

	ipc.on("window::focus", on_window_focus)
	ipc.main() # enter command loop
	```
4. In another tab of terminal execute `rofi -show run`. A window should popup, dismiss it with <kbd>Escape</kbd>
5. Look at the output of the python script

# Expected

Two "focused" lines — the first one for rofi window, the second one for terminal after rofi was dismissed.

# Actual

Just one line: for terminal after rofi was dismissed.

**Comment by RyanDwyer** [comment]
Rofi's window is an xwayland unmanaged window. This means we kind of leave it to do its own thing - we don't decorate it with borders, we don't allow standard container operations on it and it's exempt from IPC focus events. It would be interesting to see what i3 does in this case.

Rofi probably has its own transparency settings which you could tweak. If not, another option would be to replace rofi with a terminal and fzf.

**Comment by RyanDwyer** [comment]
We should probably emit an IPC event that sets the focus to `NULL`, so that you can adjust opacity on the previous view again.

**Comment by RyanDwyer** [comment]
Yeah, it's not ideal. It's even more difficult than I thought because we don't actually unfocus the view behind rofi. It retains focused styling and we haven't sent unfocus to the view.

We should check i3 and see what it does with focus styling and IPC events, then do the same.

**Comment by Hi-Angel** [comment]
> We should check i3 and see what it does with focus styling and IPC events, then do the same.

Well, I'm in office right now using i3 + Compton. When I call `rofi`, the rofi window becomes transparent, whilst the window behind it isn't. I guess that means that i3 doesn't focus on `rofi` window. But I can't say it feels right, because from a user perspective `rofi` window is clearly a focused one — it grabs the keyboard input, why wouldn't it be?

**Comment by tgerbier** [comment]
thanks Hi-Angel.
I made an adaptation avoiding global (without any particular reason), using argparse to set optional value for focused and unfocused and to handle a custom mode "screenshot" to not have transparency when using it:

`#!/usr/bin/python

# deeply inspired by https://github.com/swaywm/sway/issues/2859
# This script requires i3ipc-python package (install it from a system package manager
# or pip).
# It makes inactive windows transparent. Use `transparency_val` variable to control
# transparency strength in range of 0…1.

import i3ipc
import argparse

parser = argparse.ArgumentParser()
parser.add_argument(
        '-f', help='focused window opacity value between 0 and 1',
        type=str,
        dest='focused',
        default='1'
        )
parser.add_argument(
        '-u', help='unfocused window opacity value between 0 and 1',
        type=str,
        dest='unfocused',
        default='0.5'
        )
args = parser.parse_args()
ipc              = i3ipc.Connection()


class Last_mode:
    value = None


class Last_focused:
    window = None


class Opacity:
    value = args.unfocused


def apply_opacity(window, o):
    window.command('opacity {}'.format(o))

def windows_apply(ipc):
    for window in ipc.get_tree():
        if window.focused: 
            apply_opacity(window, args.focused)
            Last_focused.window = window
        else: apply_opacity(window, Opacity.value)

def on_window_focus(ipc, focused):
    print(focused.container.id, Opacity.value, args.focused, args.unfocused)
    if focused.container.id != Last_focused.window.id:
        windows_apply(ipc)
        Last_focused.window = focused.container

def on_mode(ipc, data):
    if data.change == "screenshot":
        Opacity.value = args.focused
        windows_apply(ipc)
    elif Last_mode == "screenshot":
        Opacity.value = args.unfocused
        windows_apply(ipc)
    else:
        Opacity.value = args.unfocused


windows_apply(ipc)
ipc.on("window::focus", on_window_focus)
ipc.on("mode", on_mode)
ipc.main()
`

Hope it have some interest for someone.
